ios/brave-ios/Sources/BraveWallet/Extensions/AssetRatioServiceExtensions.swift

```diff
@@ -7,140 +7,59 @@ import BraveCore
-          return PricesResult(
-            prior.assetPrices + new.assetPrices,
-            prior.failureCount + new.failureCount
-          )
-        }
-      )
-    }
+    return prices
   }
 }
 
-extension BraveWalletAssetRatioService {
-  /// Fetch the price for a list of assets to a list of assets over a given timeframe.
-  /// If the main call for all asset prices fails, will fetch the price for each asset individually, so one failure will not result in a failure to fetch all assets.
-  func priceWithIndividualRetry(
-    _ fromAssets: [String],
-    toAssets: [String],
-    timeframe: BraveWallet.AssetPriceTimeframe,
-    completion: @escaping (PricesResult) -> Void
-  ) {
-    let uniquePriceIds = uniqueNonEmptyPriceIds(fromAssets)
+extension Array where Element == BraveWallet.AssetPrice {
+  /// Get token price from a list of asset prices
+  /// Returns the matching AssetPrice if found, nil otherwise
+  func getTokenPrice(
+    for token: BraveWallet.BlockchainToken
+  ) -> BraveWallet.AssetPrice? {
+    guard !isEmpty else { return nil }
 
-    price(fromAssets: uniquePriceIds, toAssets: toAssets, timeframe: timeframe) {
-      [self] success, prices in
-      guard success else {
-        self.priceIndividually(
-          uniquePriceIds,
-          toAssets: toAssets,
-          timeframe: timeframe,
-          completion: completion
-        )
-        return
+    // For SOL and ETH, match by coinType, chainId, and contractAddress
+    if [BraveWallet.CoinType.sol, BraveWallet.CoinType.eth].contains(token.coin) {
+      return first { assetPrice in
+        assetPrice.coinType == token.coin
+          && assetPrice.chainId == token.chainId
+          && assetPrice.address.lowercased() == token.contractAddress.lowercased()
       }
-      completion(PricesResult(prices, 0))
     }
+
+    // For other coins, match by coinType only
+    return first { $0.coinType == token.coin }
   }
 
-  /// Fetch the price for each asset in `fromAssets` individually, so one failure will not result in a failure to fetch all assets.
-  private func priceIndividually(
-    _ fromAssets: [String],
-    toAssets: [String],
-    timeframe: BraveWallet.AssetPriceTimeframe,
-    completion: @escaping (PricesResult) -> Void
-  ) {
-    var pricesResults: [PricesResult] = []
-    let dispatchGroup = DispatchGroup()
-    fromAssets.forEach { asset in
-      dispatchGroup.enter()
-      self.price(fromAssets: [asset], toAssets: toAssets, timeframe: timeframe) { success, prices in
-        defer { dispatchGroup.leave() }
-        pricesResults.append(PricesResult(prices, success ? 0 : 1))
+  /// Updates the existing array with new AssetPrice objects, ensuring no duplicates
+  /// Two AssetPrices are considered duplicates if they have the same coinType, chainId, and address
+  mutating func update(with newPrices: [BraveWallet.AssetPrice]) {
+    for newPrice in newPrices {
+      // Find index of existing price with same coinType, chainId, and address
```

Comment:
do we not need to compare differently based on coin type like the `getTokenPrice`? i think we can update `getTokenPrice` to take `coinType`, `chainId` and `address` so that this function can use as well?